pkg/destroy/gcp/network.go

```diff
@@ -0,0 +1,647 @@
+	return nil
+}
+
+func (o *ClusterUninstaller) listAddresses() ([]string, error) {
+	o.Logger.Debugf("Listing addresses")
+	result := []string{}
+	ctx, cancel := o.contextWithTimeout()
+	defer cancel()
+	req := o.computeSvc.Addresses.List(o.ProjectID, o.Region).Fields("items(name)").Filter(o.clusterIDFilter())
+	err := req.Pages(ctx, func(list *compute.AddressList) error {
+		for _, address := range list.Items {
+			o.Logger.Debugf("Found address: %s", address.Name)
+			result = append(result, address.Name)
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to list addresses")
+	}
+	return result, nil
+}
+
+func (o *ClusterUninstaller) deleteAddress(name string) error {
+	o.Logger.Debugf("Deleting address %s", name)
+	ctx, cancel := o.contextWithTimeout()
+	defer cancel()
+	_, err := o.computeSvc.Addresses.Delete(o.ProjectID, o.Region, name).Context(ctx).Do()
+	if err != nil && !isNoOp(err) {
+		return errors.Wrapf(err, "failed to delete address %s", name)
+	}
+	return nil
+}
+
+func (o *ClusterUninstaller) destroyAddresses() error {
+	addresses, err := o.listAddresses()
+	if err != nil {
+		return err
+	}
+	for _, address := range addresses {
+		err = o.deleteAddress(address)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (o *ClusterUninstaller) listForwardingRules() ([]string, error) {
+	o.Logger.Debugf("Listing forwarding rules")
+	result := []string{}
+	ctx, cancel := o.contextWithTimeout()
+	defer cancel()
+	req := o.computeSvc.ForwardingRules.List(o.ProjectID, o.Region).Fields("items(name)").Filter(o.clusterIDFilter())
+	err := req.Pages(ctx, func(list *compute.ForwardingRuleList) error {
+		for _, forwardingRule := range list.Items {
+			o.Logger.Debugf("Found forwarding rule: %s", forwardingRule.Name)
+			result = append(result, forwardingRule.Name)
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to list forwarding rules")
+	}
+	return result, nil
+}
+
+func (o *ClusterUninstaller) deleteForwardingRule(name string) error {
+	o.Logger.Debugf("Deleting forwarding rule %s", name)
+	ctx, cancel := o.contextWithTimeout()
+	defer cancel()
+	_, err := o.computeSvc.ForwardingRules.Delete(o.ProjectID, o.Region, name).Context(ctx).Do()
+	if err != nil && !isNoOp(err) {
+		return errors.Wrapf(err, "failed to delete forwarding rule %s", name)
+	}
+	return nil
+}
+
+func (o *ClusterUninstaller) destroyForwardingRules() error {
+	forwardingRules, err := o.listForwardingRules()
+	if err != nil {
+		return err
+	}
+	for _, forwardingRule := range forwardingRules {
+		err = o.deleteForwardingRule(forwardingRule)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (o *ClusterUninstaller) listClusterBackendServices() ([]string, error) {
+	return o.listBackendServices("items(name)", o.clusterIDFilter(), nil)
+}
+
+func (o *ClusterUninstaller) listBackendServices(fields string, filter string, filterFunc func(*compute.BackendService) bool) ([]string, error) {
```

Comment:
what's the advantage of fields-filter.

Reply:
This is used for 1) listing regular backend services we create with the installer and operators for which we only need a filter string with the cluster infra ID prefix and 2) listing backend services that reference specific instance groups that we determine belong to our cluster and need to be removed (these are backend services created by the kube cloud controller)